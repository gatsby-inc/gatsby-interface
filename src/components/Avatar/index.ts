import { AvatarSize } from "./types"
import { AvatarProps } from "./Avatar"
import { AvatarsGroupProps, AvatarDescriptor } from "./AvatarsGroup"

export type AvatarSize = AvatarSize
export type AvatarProps = AvatarProps
export type AvatarsGroupProps = AvatarsGroupProps
export type AvatarDescriptor = AvatarDescriptor

export { default as Avatar } from "./Avatar"
export { default as AvatarsGroup } from "./AvatarsGroup"
