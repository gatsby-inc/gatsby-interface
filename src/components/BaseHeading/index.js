export { default as BaseHeading } from "./BaseHeading"
