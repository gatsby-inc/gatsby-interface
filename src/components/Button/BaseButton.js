import React from "react"
import styled from "@emotion/styled"
import { keyframes } from "@emotion/core"

import { fontFamilies, radius, palette } from "../../utils/presets"

import ButtonSkeleton, {
  buttonPropTypes,
  buttonDefaultPropTypes,
  SkeletonStyledComponent,
} from "./Button.Skeleton"

const loading = keyframes`
  0% {
    transform: translateX(0.2em) rotate(0);
  }
  100% {
    transform: translateX(0.2em) rotate(360deg);
  }
`

const pulse = keyframes`
  33% {
    transform: translateX(0.2em) scale(1.05);
    }
  66% {
    transform: translateX(0.2em) scale(0.85);
  }

`

const sizes = {
  S: {
    fontSize: 0.875,
    height: 1.5,
    horizontalPadding: 0.5,
    verticalPadding: 0.3,
  },
  M: {
    fontSize: 1,
    height: 2,
    horizontalPadding: 0.75,
    verticalPadding: 0.45,
  },
  L: {
    fontSize: 1.125,
    height: 2.25,
    horizontalPadding: 1,
    verticalPadding: 0.55,
  },
  XL: {
    fontSize: 1.5,
    height: 3.25,
    horizontalPadding: 1.25,
    verticalPadding: 0.65,
  },
}

export const StyledBaseButton = styled(SkeletonStyledComponent)`
  align-items: center;
  border: ${palette.grey[600]};
  border-radius: ${radius.default};
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  font-family: ${fontFamilies.headerFontFamily};
  font-size: ${props => sizes[props.size].fontSize}rem;
  justify-content: center;
  line-height: 1;
  min-height: ${props => sizes[props.size].height}rem;
  padding: ${props => sizes[props.size].verticalPadding}rem
    ${props => sizes[props.size].horizontalPadding}rem;
  text-decoration: none;
  transition: 0.5s;

  &[disabled],
  &[disabled]:hover {
    cursor: not-allowed;
    opacity: 0.5;
  }

  svg {
    animation-name: ${props => (props.showLoading ? loading : ``)};
    animation-duration: 1s;
    animation-iteration-count: infinite;
    animation-timing-function: linear;
    flex-shrink: 0;
    transform: translateX(0.2em) scale(1);
  }

  &:hover:not([disabled]) {
    svg {
      animation: ${pulse} 1s linear infinite;
    }
  }
`

const BaseButton = props => (
  <ButtonSkeleton StyledComponent={StyledBaseButton} {...props} />
)

BaseButton.propTypes = buttonPropTypes
BaseButton.defaultProps = buttonDefaultPropTypes

export default BaseButton
