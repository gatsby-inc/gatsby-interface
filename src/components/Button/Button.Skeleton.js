import React, { Fragment } from "react"
import PropTypes from "prop-types"
import styled from "@emotion/styled"
import { MdRefresh } from "react-icons/md"
import { Link } from "gatsby"

export const SkeletonStyledComponent = styled(`button`)``

export const buttonPropTypes = {
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  loadingLabel: PropTypes.string,
  href: PropTypes.string,
  size: PropTypes.string,
  to: PropTypes.string,
}

const buttonSkeletonPropTypes = {
  ...buttonPropTypes,
  StyledComponent: PropTypes.any,
}

export const buttonDefaultPropTypes = {
  disabled: false,
  loading: false,
  loadingLabel: `Loading`,
  size: `L`,
}

export const buttonSkeletonDefaultPropTypes = {
  ...buttonDefaultPropTypes,
  StyledComponent: SkeletonStyledComponent,
}

const ButtonSkeleton = ({
  StyledComponent,
  children,
  disabled,
  href,
  loading,
  loadingLabel,
  to,
  ...rest
}) => {
  if (href) {
    const ComponentAsExternalLink = StyledComponent.withComponent(`a`)
    return (
      <ComponentAsExternalLink href={href} {...rest}>
        {children}
      </ComponentAsExternalLink>
    )
  }

  if (to) {
    const ComponentAsInternalLink = StyledComponent.withComponent(Link)
    return (
      <ComponentAsInternalLink to={to} {...rest}>
        {children}
      </ComponentAsInternalLink>
    )
  }

  return (
    <StyledComponent disabled={disabled} showLoading={loading} {...rest}>
      {loading ? (
        <Fragment>
          {loadingLabel} <MdRefresh />
        </Fragment>
      ) : (
        children
      )}
    </StyledComponent>
  )
}

ButtonSkeleton.propTypes = buttonSkeletonPropTypes
ButtonSkeleton.defaultProps = buttonSkeletonDefaultPropTypes

export default ButtonSkeleton
