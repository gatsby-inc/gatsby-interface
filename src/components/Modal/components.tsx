import { DialogContent } from "@reach/dialog"
import styled from "@emotion/styled"

export const Content = styled(DialogContent)`
  outline: none;
  margin: 0;
  padding: 0;
  background: transparent;
  width: auto;
`
