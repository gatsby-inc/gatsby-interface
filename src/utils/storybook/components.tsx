/** @jsx jsx */
import { jsx } from "@emotion/core"
import React from "react"
import { Heading } from "../../components/Heading"
import { Theme } from "../../theme"

export type StoryPropVariantProps = React.ComponentPropsWithoutRef<"div"> & {
  propName: string
  propValue: any
  displayValueInQuotes?: boolean
}

export function StoryPropVariant({
  propName,
  propValue,
  displayValueInQuotes = true,
  children,
  ...rest
}: StoryPropVariantProps) {
  return (
    <div {...rest}>
      <Heading css={(theme: Theme) => ({ marginBottom: theme.space[4] })}>
        <span
          css={(theme: Theme) => ({
            fontFamily: theme.fonts.monospace,
            fontWeight: theme.fontWeights.body,
            fontSize: theme.fontSizes[1],
            backgroundColor: theme.colors.grey[10],
            padding: theme.space[2],
          })}
        >
          {propName}=
          {displayValueInQuotes ? `"${propValue}"` : `{${propValue}}`}
        </span>
      </Heading>
      {children}
    </div>
  )
}
