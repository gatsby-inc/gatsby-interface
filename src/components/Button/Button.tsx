/** @jsx jsx */
import { jsx, css } from "@emotion/core"
import React from "react"
import { MdRefresh } from "react-icons/md"

import { BaseButton, BaseButtonProps } from "../BaseButton"
import styles from "../../theme/styles/button"

export type ButtonSize = "XL" | "L" | "M" | "S"
export type ButtonTone = "BRAND" | "SUCCESS" | "DANGER" | "NEUTRAL"
export type ButtonVariant = "PRIMARY" | "SECONDARY" | "GHOST"

export type ButtonStyleProps = {
  size?: ButtonSize
  tone?: ButtonTone
  variant?: ButtonVariant
  leftIcon?: React.ReactNode
  rightIcon?: React.ReactNode
}

export type ButtonProps = BaseButtonProps & ButtonStyleProps

export function getButtonStyles({
  children,
  size = `L`,
  tone = `BRAND`,
  variant = `PRIMARY`,
  loading,
  leftIcon,
  rightIcon,
}: {
  children: React.ReactNode
  loading?: boolean
} & ButtonStyleProps): {
  css: ReturnType<typeof css>
  children: React.ReactNode
} {
  return {
    css: [
      styles.base({ loading, leftIcon, rightIcon }),
      styles.sizes[size],
      styles.variants[variant]({ tone }),
    ] as any,
    children:
      leftIcon || rightIcon ? (
        <React.Fragment>
          {leftIcon}
          {children}
          {rightIcon}
        </React.Fragment>
      ) : (
        children
      ),
  }
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const {
      children,
      loading,
      LoadingIcon = MdRefresh,
      size,
      tone,
      variant,
      leftIcon,
      rightIcon,
      ...rest
    } = props

    return (
      <BaseButton
        {...getButtonStyles({
          children,
          loading,
          size,
          tone,
          variant,
          leftIcon,
          rightIcon,
        })}
        loading={loading}
        LoadingIcon={LoadingIcon}
        {...rest}
        ref={ref}
      />
    )
  }
)
