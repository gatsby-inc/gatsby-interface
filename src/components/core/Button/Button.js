/** @jsx jsx */
import { jsx } from "@emotion/core"
import PropTypes from "prop-types"
import { MdRefresh, MdArrowForward } from "react-icons/md"
import { BaseButton } from "../../skeletons/BaseButton"
import styles from "../../../theme/styles/button"

const Button = props => {
  const {
    children,
    loading,
    LoadingIcon = MdRefresh,
    size = `L`,
    tone = `BRAND`,
    variant = `PRIMARY`,
    ...rest
  } = props
  const DefaultIcon = variant === `PRIMARY` && MdArrowForward

  return (
    <BaseButton
      css={{
        ...styles.base({ loading }),
        ...styles.sizes[size],
        ...styles.variants[variant]({ tone }),
      }}
      DefaultIcon={DefaultIcon}
      loading={loading}
      LoadingIcon={LoadingIcon}
      {...rest}
    >
      {children}
    </BaseButton>
  )
}

Button.propTypes = {
  size: PropTypes.oneOf([`L`, `M`, `XL`, `S`]),
}

export default Button
