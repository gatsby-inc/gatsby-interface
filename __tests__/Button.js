import React from "react"
import { render } from "@testing-library/react"

import { PrimaryButton } from "../src/components/Button"

describe(`<PrimaryButton>`, () => {
  test(`renders unchanged`, async () => {
    const { container } = render(<PrimaryButton>Click me!</PrimaryButton>)

    expect(container).toMatchSnapshot()
  })
})
