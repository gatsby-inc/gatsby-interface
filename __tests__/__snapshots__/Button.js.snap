// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrimaryButton> renders unchanged 1`] = `
@keyframes animation-0 {
  33% {
    -webkit-transform: translateX(0.2em) scale(1.05);
    -ms-transform: translateX(0.2em) scale(1.05);
    transform: translateX(0.2em) scale(1.05);
  }

  66% {
    -webkit-transform: translateX(0.2em) scale(0.85);
    -ms-transform: translateX(0.2em) scale(0.85);
    transform: translateX(0.2em) scale(0.85);
  }
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: #635e69;
  border-radius: 4px;
  box-sizing: border-box;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: Futura PT,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen,Ubuntu,Cantarell,Fira Sans,Droid Sans,Helvetica Neue,Arial,sans-serif;
  font-size: 1.125rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  min-height: 2.25rem;
  padding: 0.55rem 1rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.5s;
  transition: 0.5s;
  background: #663399;
  border: 0;
  color: #ffffff;
  font-weight: bold;
}

.emotion-0[disabled],
.emotion-0[disabled]:hover {
  cursor: not-allowed;
  opacity: 0.5;
}

.emotion-0 svg {
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transform: translateX(0.2em) scale(1);
  -ms-transform: translateX(0.2em) scale(1);
  transform: translateX(0.2em) scale(1);
}

.emotion-0:hover:not([disabled]) svg {
  -webkit-animation: animation-0 1s linear infinite;
  animation: animation-0 1s linear infinite;
}

.emotion-0:focus,
.emotion-0:hover {
  background: #542c85;
}

<div>
  <button
    class="emotion-0 emotion-1"
  >
    Click me!
  </button>
</div>
`;
